---
title: "Exploratory analyses of the pilot data from the Promort study"
subtitle: "Quality assessment assessment of the RNA-seq data"
author: "`r paste0('Ashfaq Ali â€¢ ',format(Sys.time(), '%d-%b-%Y'))`"
output:
  bookdown::html_document2:
          toc: true
          toc_float: true
          toc_depth: 4
          number_sections: true
          theme: flatly
          highlight: tango
          df_print: paged
          code_folding: "none"
          self_contained: false
          keep_md: false
          encoding: 'UTF-8'
          css: "assets/report.css"
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/cob-aaf/Documents/Projects/PP_4730/NPD_updated/reports/')
```


```{r echo=FALSE, message=FALSE, warning=FALSE, include=FALSE}
library(tidyverse)
library(Hmisc)
library(tximport)
#library(tximportData)
library(DESeq2)
library(naniar)
library(limma)
library(Glimma)
library(edgeR)
library(ReportingTools)
library(BiocParallel)
library(readxl) # read from and write to excel files
library(biomaRt)
library(beanplot)
library(reshape2)
library(ggpubr)
```
---



## Exploratory analyses

Due to large shift in foldchange shifting with mean counts, the potential possibility of batch effect was 

 **Read counts per sample and genes to explore the quality of libraries**
 
 - A) Raw counts violin plots for number of reads per samples
 - B) Genes with more than 5 reads per sample
 - C) Total raw counts per sample
 - D) PCA 


```{r, eval=FALSE, echo=FALSE}
#melt(DT, id=c("i_1", "i_2"), measure=3)
#wide_data<-cbind(colData(se), t(assays(se)$counts+1))
long_data <- melt((assays(se)$counts)) 
long_data2<- inner_join(long_data, phenotypes, by =c("Var2"="sample_id"))
long_data2$countsums5 <- as.numeric(long_data2$value>5)
```

```{r , eval = FALSE, echo = FALSE, fig.width=12, fig.height=15, warning = FALSE, message= FALSE}
p <- ggplot(long_data2, aes(x = Var2, y = log10(value), color = npd)) + geom_violin() +
stat_summary(geom = 'point', fun.y = median) + 
 labs(title = "A) Sample wise Read counts", x = "Sample", y = expression('Log'[10]~'read counts')) +
#p + labs(title = "New plot title", subtitle = "A subtitle", tag = "A") + labs(title = NULL)
labs(caption = "(based on count data without filtering transcripts or samples)") +
theme( axis.line = element_line(colour = "darkblue",  size = 1, linetype = 1), axis.text.x = element_text(angle = 90)) +
scale_x_discrete(labels=se$sample_id) +
theme(axis.text.x = element_text(angle = 90)) 
#+ geom_jitter(shape=16, position=position_jitter(0.2))
#png("/Users/cob-aaf/Documents/Projects/PP_4730/NPD_updated/reports/Figures/violin_rawcounts.png", width = 3024, height = 2000, res = 300)

g <- ggplot(as.data.frame(colData(se)), aes(y=countsums5, x=sample_id, color = npd)) +
# Number of cars in each class:
geom_bar(stat="identity") + 
  labs(title = "B) Number of expressed gene ncount > 5", x = "Sample", y = expression('Number of genes')) +
#p + labs(title = "New plot title", subtitle = "A subtitle", tag = "A") + labs(title = NULL)
labs(caption = "(based on count data without filtering transcripts or samples)") +
  theme(axis.text.x = element_text(angle = 90)) 

# Total engine displacement of each class
#png("/Users/cob-aaf/Documents/Projects/PP_4730/NPD_updated/reports/Figures/bar_rawcounts_pergene.png", width = 3400, height = 1600, res = 300)

se$sumst <- assays(se)$counts %>% colSums()

s <- ggplot(as.data.frame(colData(se)), aes(y=sumst, x=sample_id, color = npd)) +
# Number of cars in each class:
geom_bar(stat="identity") + 
  labs(title = "C) Total counts (raw)", x = "Sample", y = expression('Number of genes')) +
#p + labs(title = "New plot title", subtitle = "A subtitle", tag = "A") + labs(title = NULL)
labs(caption = "(based on count data without filtering transcripts or samples)") +
  theme(axis.text.x = element_text(angle = 90)) 

# Total engine displacement of each class
#png("/Users/cob-aaf/Documents/Projects/PP_4730/NPD_updated/reports/Figures/bar_rawcounts_pergene.png", width = 3400, height = 1600, res = 300)
ggpubr::ggarrange(p, g, s, nrow = 3) %>% ggexport(filename = "reports/Figures/bars.png", res = 300, height = 1200, width = 1600)


#dev.off()
```

The Violoin plot (A), and bar plots indicate that the coverage for geneome wide tracriptomics analyses may not be appropriate as there is significont variation in the sequencing depth. Normalizing based on size factor (estimates for relative library size) may not be sufficient to normalize the data. 

---

### PCA analyses

PCA plot shows explaind variance in the data by PCs.

```{r, eval = FALSE, echo = FALSE, fig.width=10, fig.height=18, warning = FALSE, message= FALSE}
vsndds <- vst(dds, blind=FALSE)

pcaobj <- prcomp(x=t(assays(vsndds)[[1]]))

percentVar <- pcaobj$sdev^2/sum(pcaobj$sdev^2)

pcamat1 <- as.data.frame(pcaobj$x)

pcamat2 <- as.data.frame(merge(pcamat1,colData(vsndds),by=0))
```

```{r, eval=FALSE}
png("reports/Figures/PCA_bar.png", units = "cm", height = 10, width = 8, res = 1500)
barplot(percentVar[1:10]*100, ylab = "Percent varinace explained", xlab = "PC 1 to 10"  ,main = "Percent variace explained", col = "purple")
dev.off()
```
[PCA_Barplot](reports/Figures/PCA_bar.png)

```{r, eval = FALSE, echo = FALSE, fig.width=10, fig.height=6, warning = FALSE, message= FALSE}
# The basic PCA plot for NPD
p1 <- ggplot(pcamat2,aes(x = PC1, y = PC2, colour=ind, label = sample_id )) +
  
  geom_point(size = 1) +
  geom_text(size=2,angle = 30, nudge_y = 5, nudge_x = -4)+
  geom_jitter(width = 1)+
  #geom_label(label.size = 0.15, nudge_y = 1.4, label.padding = unit(0.15, "lines"))+
  xlab(paste0("PC1: ", round(percentVar[1] * 
        100), "% variance")) + ylab(paste0("PC2: ", round(percentVar[2] * 
        100), "% variance")) + coord_fixed() + theme_bw(base_size = 11)+theme(legend.position = "top",
          legend.direction = "horizontal")
 png("reports/Figures/PCA_npd_ind.png", units = "cm", height = 12, width = 15, res = 1500)
  p1
dev.off()
```

[PCA_npd](reports/Figures/PCA_npd_ind.png)

```{r, eval = FALSE, echo = FALSE, fig.width=10, fig.height=6, warning = FALSE, message= FALSE}
p1 <- ggplot(pcamat2,aes(x = PC2, y = PC3, colour=sizeFactor, label = npd )) +
  
  geom_point(size = 1) +
  geom_text(size=2, nudge_y = 3, nudge_x = 2)+
  geom_jitter(width = 1)+
  #geom_label(label.size = 0.15, nudge_y = 1.4, label.padding = unit(0.15, "lines"))+
  xlab(paste0("PC3: ", round(percentVar[3] * 
        100), "% variance")) + ylab(paste0("PC2: ", round(percentVar[2] * 
        100), "% variance")) + coord_fixed() + theme_bw(base_size = 11)+theme(legend.position = "top",
          legend.direction = "horizontal")
 png("reports/Figures/PCA_npd_sizeF_pc23.png", units = "cm", height = 12, width = 15, res = 1500)
  p1
dev.off()
```

[PCA_npd](reports/Figures/PCA_npd_sizeF_pc23.png)

The PCA plots indicate that there is grouping on size factor and D samples tend to have larger sequencing depth either due to either bias at the sample preparation and sequencing or due to bilogical differences in the tissue types. The P samples present a few outliers and tend to have smaller libraries, the seqiuencing data is noisy compared with N and D samples.

---

## Filter on counts and dispersion 

The gene dispersion estimates based on Peter's filtering indicate issues with sequencing baises and many genes appear to have variance equal to mean (indicated by dispersion values close to 1). Further filtering to select genes with higher expression (scaled TPM > 10) and across many samples (min 40) seems to help remove the genes with biased estimaes. The dispersion plot improves significantly after the filtering. 
Six of the outlier samples clearly visible in the PCA plots were also removed.

```{r, eval=FALSE, echo=FALSE}
sums<- colSums(assays(se)$counts>5)
se$countsums5<-colSums(assays(se)$counts>5)
sef$countsums5<-colSums(assays(sef)$counts>5)
```

```{r, eval = FALSE, echo = FALSE, fig.width=10, fig.height=6, warning = FALSE, message= FALSE}
keepgenes <- rowSums(assays(se)$lenScaledTPM>10) >=40
out_samples<- pcamat2[pcamat2$PC1<(-110), "sample_id"]
sef<- se[keepgenes,colnames(se) %nin% out_samples]
sefdds<- DESeqDataSet(sef, design = ~ind+npd)
sefdds$npd <- relevel(sefdds$npd, ref = "N")
sefdds$ind <- factor(sefdds$ind)
#keepgenes <- rowVars(assays(se)$counts) >=40
#sef<-se[keepgenes,]
dim(sefdds)
register(MulticoreParam(8))
sefdds<-estimateSizeFactors(sefdds)
sefdds <- DESeq2::estimateDispersions(sefdds)

sefdds <-DESeq(sefdds, parallel = TRUE)

#sefdds <- results(sefdds,  contrast = c("npd","D","P"), alpha = 0.05) 

fil_d_vs_n <- results(sefdds,  contrast = c("npd","N","D"), alpha = 0.05) 

png("reports/Figures/Filteres_d_vs_n_allsamples_MA).png", units = "cm", height = 12, width = 15, res = 1500)
  results(sefdds,  contrast = c("npd","N","D"), alpha = 0.05) %>% DESeq2::plotMA()
dev.off()

fil_d_vs_nlfc <- lfcShrink(sefdds, contrast = c("npd","N","D"), res = fil_d_vs_n, parallel = TRUE )

fil_d_vs_n <- fil_d_vs_n %>% as.data.frame() %>% rownames_to_column(., var = "ensembl_gene_id") %>% arrange(padj)
fil_d_vs_nlfc <- fil_d_vs_nlfc %>% as.data.frame() %>% rownames_to_column(., var = "ensembl_gene_id") %>% arrange(padj)

fil_d_vs_n <- fil_d_vs_n %>% mutate(sig=padj < 0.05)
fil_d_vs_nlfc <- fil_d_vs_nlfc %>% mutate(sig=padj < 0.05)
table(fil_d_vs_n$sig)

png("reports/Figures/MApv_d_vs_n_gg_LFSfilt.png", units = "cm", height = 12, width = 15, res = 1500)
p <- ggplot(fil_d_vs_nlfc, aes(baseMean, log2FoldChange , col=sig))
p + geom_point() + scale_x_log10() + ylim(-1.55,1.5)
dev.off()


fil_p_vs_n <- results(sefdds,  contrast = c("npd","N","P"), alpha = 0.05 ) 
fil_p_vs_nlfc <- lfcShrink(sefdds, contrast = c("npd","N","P"), res = fil_p_vs_n, parallel = TRUE )

fil_p_vs_n <- fil_p_vs_n %>% as.data.frame() %>% rownames_to_column(., var = "ensembl_gene_id") %>% arrange(padj)
fil_p_vs_nlfc <- fil_p_vs_nlfc %>% as.data.frame() %>% rownames_to_column(., var = "ensembl_gene_id") %>% arrange(padj)

fil_p_vs_n <- fil_p_vs_n  %>% mutate(sig=padj < 0.05)
fil_p_vs_nlfc <- fil_p_vs_nlfc  %>% mutate(sig=padj < 0.05)

png("reports/Figures/MApv_p_vs_n_gg_filtLFS.png", units = "cm", height = 12, width = 15, res = 1500)
p <- ggplot(fil_p_vs_nlfc, aes(baseMean, log2FoldChange , col=sig))
p + geom_point() + scale_x_log10() + ylim(-1.55,1.5)
dev.off()


fil_p_vs_d <- results(sefdds,  contrast = c("npd","D","P"), alpha = 0.05) 
fil_p_vs_dlfc <- lfcShrink(sefdds, contrast = c("npd","D","P"), res = fil_p_vs_d, parallel = TRUE )

fil_p_vs_d <- fil_p_vs_d %>%  as.data.frame() %>% rownames_to_column(., var = "ensembl_gene_id") %>% arrange(padj)
fil_p_vs_dlfc <- fil_p_vs_dlfc %>%  as.data.frame() %>% rownames_to_column(., var = "ensembl_gene_id") %>% arrange(padj)
 

fil_p_vs_d <- fil_p_vs_d %>% mutate(sig=padj < 0.05)
fil_p_vs_dlfc <- fil_p_vs_dlfc %>% mutate(sig=padj < 0.05)

png("reports/Figures/MApv_p_vs_d_gg_filtLFC.png", units = "cm", height = 12, width = 15, res = 1500)
p <- ggplot(fil_p_vs_dlfc, aes(baseMean, log2FoldChange , col=sig))
p + geom_point() + scale_x_log10() + ylim(-1.55,1.5)
dev.off()


plotCounts(dds = dds,gene="ENSG00000112214",intgroup="npd",normalized=T, transform = FALSE)
plotCounts(dds = dds,gene="ENSG00000061676",intgroup="npd",normalized=T,transform = FALSE)
plotCounts(dds = dds,gene="ENSG00000198523",intgroup="npd",normalized=T,transform = FALSE)

```


```{r, eval=FALSE}
PV_NPD_fil<- list(DN = inner_join(gene_ann, fil_d_vs_n), PN = inner_join(gene_ann,  fil_p_vs_n), PD = inner_join(gene_ann, fil_p_vs_d))
PV_NPD_fil$TPM <- as.data.frame(assays(ddsf)$lenScaledTPM)  %>% rownames_to_column(., var = "ensembl_gene_id")
PV_NPD_fil$counts <- as.data.frame(assays(ddsf)$counts) %>% rownames_to_column(., var = "ensembl_gene_id")
```


```{r, eval=FALSE}
saveRDS(PV_NPD_fil, file = "../NPD_updated/NPD_Results/PV_rep_results_NPD_fil.rds")
writexl::write_xlsx(x = PV_NPD_fil, path = "../NPD_updated/NPD_Results/PV_rep_results_NPD_fil.xlsx")
```

```{r, eval=FALSE}
PV_NPD_fil_lfs<- list(DN = inner_join(gene_ann, fil_d_vs_nlfc), PN = inner_join(gene_ann,  fil_p_vs_nlfc), PD = inner_join(gene_ann, fil_p_vs_dlfc))
PV_NPD_fil_lfs$TPM <- as.data.frame(assays(ddsf)$lenScaledTPM)  %>% rownames_to_column(., var = "ensembl_gene_id")
PV_NPD_fil_lfs$counts <- as.data.frame(assays(ddsf)$counts) %>% rownames_to_column(., var = "ensembl_gene_id")
```


```{r, eval=FALSE}
writexl::write_xlsx(x = PV_NPD_fil_lfs, path = "../NPD_updated/NPD_Results/PV_rep_results_NPD_fil_lfs.xlsx")
saveRDS(PV_NPD_fil_lfs, file = "../NPD_updated/NPD_Results/PV_rep_results_NPD_fil_lfs.rds")

```